### PR TITLE
Remove writes from table

### PR DESCRIPTION
**Pull request by nnsgmsone**
## What type of PR is this?

- [ ] API-change
- [] BUG
- [ ] Improvement
- [ ] Documentation
- [] Feature
- [ ] Test and CI
- [X] Code Refactoring

## Which issue(s) this PR fixes:

issue #14552

## What this PR does / why we need it:
unnecessary caching of write offset is the root of all evils.

### CODE REVIEW

**Review by fengttt — COMMENTED** [review]
I will be damned if this one passes bvt.

**Comment by nnsgmsone** [comment]
The current scenario involves SQL executions where not all statement IDs are consistently added at the beginning. Consequently, attempting to retrieve a snapshot by statement ID during specified ranges may lead to inaccurate data retrieval. This issue risks losing the writes from the preceding SQL queries.

**Review by triump2020 — APPROVED** [review]

File: pkg/vm/engine/disttae/txn_table.go [hunk 1]
```diff
@@ -2273,3 +2267,16 @@ func (tbl *txnTable) readNewRowid(vec *vector.Vector, row int,
...
 	return plan2.BindFuncExprImplByPlanExpr(tbl.proc.Load().Ctx, "=", []*plan.Expr{pkExpr, constExpr})
 }
+
+func (tbl *txnTable) getSnapshotWriteOffset() int {
+	tbl.Lock()
+	defer tbl.Unlock()
+	return tbl.snapshotWriteOffset
+}
+
+// Before executing an SQL statement, take a snapshot to ensure that the SQL statement does not read the data it writes.
```
Comment (fengttt) [inline pkg/vm/engine/disttae/txn_table.go:2277]: why do you need this -- we have just got rid of the write offset in table, why you want it back?

File: pkg/vm/engine/disttae/txn.go [hunk 2]
```diff
@@ -1025,11 +1025,8 @@ func (txn *Transaction) clearTableCache() {
...
 }
 
-func (txn *Transaction) getWriteOffset() int {
-	if txn.statementID > 0 {
-		txn.Lock()
-		defer txn.Unlock()
-		return txn.statements[txn.statementID-1]
-	}
-	return 0
+func (txn *Transaction) snapshotWriteOffset() int {
```
Comment (fengttt) [inline pkg/vm/engine/disttae/txn.go:1028]: Len(txn.writes) is not write offset of a snapshot in the middle of transaction.

**Review by zhangxu19830126 — APPROVED** [review]

**Review by daviszhen — APPROVED** [review]

**Review by fengttt — APPROVED** [review]

**Pull request merged**